File: zettelgarden-mcp/handlers/calendar.py
"""
Calendar handlers for Zettelgarden MCP Server.
"""

import httpx
from datetime import datetime, timedelta
from typing import Any


async def list_external_calendars(client: httpx.AsyncClient) -> str:
    """List all external calendar subscriptions for the current user."""
    try:
        response = await client.get("/user/external-calendars")
        response.raise_for_status()
        data = response.json()

        if not data:
            return "No external calendars found."

        calendars = data
        result = f"Found {len(calendars)} external calendar(s):\n\n"

        for cal in calendars:
            result += f"- **{cal['name']}** (ID: {cal['id']})\n"
            result += f"  URL: {cal['url']}\n"
            result += f"  Sync: {'enabled' if cal['sync_enabled'] else 'disabled'}\n"
            result += f"  Color: {cal['color']}\n"
            if cal.get('last_synced_at'):
                result += f"  Last synced: {cal['last_synced_at']}\n"
            if cal.get('last_error'):
                result += f"  Last error: {cal['last_error']}\n"
            result += "\n"

        return result.strip()
    except httpx.HTTPError as e:
        return f"Error listing external calendars: {e}"
    except Exception as e:
        return f"Unexpected error: {e}"


async def list_external_events(client: httpx.AsyncClient, args: dict[str, Any]) -> str:
    """List external calendar events within a date range."""
    start = args.get("start")
    end = args.get("end")

    if not start or not end:
        return "Error: Both 'start' and 'end' dates are required (ISO 8601 format)."

    try:
        # Validate date format
        try:
            datetime.fromisoformat(start.replace('Z', '+00:00'))
            datetime.fromisoformat(end.replace('Z', '+00:00'))
        except ValueError:
            return "Error: Invalid date format. Use ISO 8601 format (e.g., '2026-01-01T00:00:00Z')."

        response = await client.get(
            "/user/external-events",
            params={"start": start, "end": end}
        )
        response.raise_for_status()
        data = response.json()

        events = data.get("events", [])
        total = data.get("total", 0)

        if not events:
            return f"No external events found between {start} and {end}."

        result = f"Found {total} external event(s):\n\n"

        for event in events:
            start_time = datetime.fromisoformat(event['start_time'].replace('Z', '+00:00'))
            result += f"- **{event['title']}** (ID: {event['id']})\n"
            result += f"  Time: {start_time.strftime('%Y-%m-%d %H:%M')}"
            if event.get('all_day'):
                result += " (all day)"
            result += "\n"
            if event.get('location'):
                result += f"  Location: {event['location']}\n"
            if event.get('card_pk') and event['card_pk'] > 0:
                result += f"  Linked to card: {event['card_pk']}\n"
            if event.get('description'):
                desc = event['description'][:100] + "..." if len(event['description']) > 100 else event['description']
                result += f"  Description: {desc}\n"
            result += "\n"

        return result.strip()
    except httpx.HTTPError as e:
        return f"Error listing external events: {e}"
    except Exception as e:
        return f"Unexpected error: {e}"


async def link_event_to_card(client: httpx.AsyncClient, args: dict[str, Any]) -> str:
    """Link an external calendar event to a card."""
    event_id = args.get("event_id")
    card_pk = args.get("card_pk")

    if event_id is None or card_pk is None:
        return "Error: Both 'event_id' and 'card_pk' are required."

    try:
        response = await client.put(
            f"/user/external-events/{event_id}/link",
            json={"card_pk": card_pk}
        )
        response.raise_for_status()
        event = response.json()

        return f"Successfully linked event '{event['title']}' to card {card_pk}."
    except httpx.HTTPStatusError as e:
        if e.response.status_code == 404:
            return "Error: Event or card not found."
        elif e.response.status_code == 400:
            return f"Error: {e.response.json().get('error', 'Invalid request')}"
        else:
            return f"Error linking event to card: {e}"
    except httpx.HTTPError as e:
        return f"Error linking event to card: {e}"
    except Exception as e:
        return f"Unexpected error: {e}"

File: go-backend/services/tools/calendar/calendar.go
// Package calendar provides calendar-related data access and business logic
// for the Zettelgarden tool registry.
//
// This package contains functions for managing external calendar integrations,
// including listing calendars, events, and linking events to cards.
package calendar

import (
	"database/sql"
	"fmt"
	"time"

	"go-backend/models"
)

// GetCalendars retrieves all external calendar subscriptions for a user
func GetCalendars(db *sql.DB, userID int) ([]models.ExternalCalendar, error) {
	rows, err := db.Query(`
		SELECT id, user_id, name, url, sync_enabled, sync_interval_hours,
		       color, last_synced_at, last_error, created_at, updated_at,
		       username
		FROM external_calendars
		WHERE user_id = $1
		ORDER BY created_at
	`, userID)

	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var calendars []models.ExternalCalendar
	for rows.Next() {
		var cal models.ExternalCalendar
		var username sql.NullString
		if err := rows.Scan(
			&cal.ID,
			&cal.UserID,
			&cal.Name,
			&cal.URL,
			&cal.SyncEnabled,
			&cal.SyncIntervalHours,
			&cal.Color,
			&cal.LastSyncedAt,
			&cal.LastError,
			&cal.CreatedAt,
			&cal.UpdatedAt,
			&username,
		); err != nil {
			return nil, err
		}
		if username.Valid {
			cal.Username = &username.String
		}
		calendars = append(calendars, cal)
	}

	return calendars, nil
}

// GetEventsInRange retrieves external calendar events within a date range
func GetEventsInRange(db *sql.DB, userID int, start, end time.Time, limit, offset int) ([]models.ExternalEvent, int, error) {
	// First, get the total count
	var total int
	err := db.QueryRow(`
		SELECT COUNT(*)
		FROM external_events
		WHERE user_id = $1
		  AND start_time >= $2
		  AND end_time <= $3
	`, userID, start, end).Scan(&total)

	if err != nil {
		return nil, 0, err
	}

	// Set default pagination values
	if limit <= 0 {
		limit = 100
	}
	if offset < 0 {
		offset = 0
	}

	// Get events - simplified query without JOIN for now
	query := `
		SELECT id, user_id, external_calendar_id, title, description,
		       start_time, end_time, all_day, location,
		       external_uid, external_url, recurrence_rule,
		       recurrence_id, recurrence_instance,
		       color, card_pk, created_at, updated_at, last_synced_at
		FROM external_events
		WHERE user_id = $1
		  AND start_time >= $2
		  AND end_time <= $3
		ORDER BY start_time
		LIMIT $4 OFFSET $5
	`

	rows, err := db.Query(query, userID, start, end, limit, offset)
	if err != nil {
		return nil, 0, err
	}
	defer rows.Close()

	var events []models.ExternalEvent
	for rows.Next() {
		var event models.ExternalEvent
		var description, location, externalUID, externalURL, recurrenceRule, recurrenceID, color sql.NullString
		var externalCalendarID sql.NullInt32
		var cardPK sql.NullInt32
		var recurrenceInstance sql.NullInt32
		var lastSyncedAt sql.NullTime

		if err := rows.Scan(
			&event.ID,
			&event.UserID,
			&externalCalendarID,
			&event.Title,
			&description,
			&event.StartTime,
			&event.EndTime,
			&event.AllDay,
			&location,
			&externalUID,
			&externalURL,
			&recurrenceRule,
			&recurrenceID,
			&recurrenceInstance,
			&color,
			&cardPK,
			&event.CreatedAt,
			&event.UpdatedAt,
			&lastSyncedAt,
		); err != nil {
			return nil, 0, err
		}

		// Convert nullable fields to pointers
		if externalCalendarID.Valid {
			id := int(externalCalendarID.Int32)
			event.ExternalCalendarID = &id
		}
		if description.Valid {
			event.Description = &description.String
		}
		if location.Valid {
			event.Location = &location.String
		}
		if externalUID.Valid {
			event.ExternalUID = &externalUID.String
		}
		if externalURL.Valid {
			event.ExternalURL = &externalURL.String
		}
		if recurrenceRule.Valid {
			event.RecurrenceRule = &recurrenceRule.String
		}
		if recurrenceID.Valid {
			event.RecurrenceID = &recurrenceID.String
		}
		if recurrenceInstance.Valid {
			instance := int(recurrenceInstance.Int32)
			event.RecurrenceInstance = &instance
		}
		if color.Valid {
			event.Color = &color.String
		}
		if cardPK.Valid {
			pk := int(cardPK.Int32)
			event.CardPK = &pk
		}
		if lastSyncedAt.Valid {
			event.LastSyncedAt = &lastSyncedAt.Time
		}

		events = append(events, event)
	}

	return events, total, nil
}

// LinkEventToCard links an external calendar event to a card
func LinkEventToCard(db *sql.DB, userID int, eventID int, cardPK int) (*models.ExternalEvent, error) {
	// Verify the event belongs to the user
	var exists bool
	err := db.QueryRow(`
		SELECT EXISTS(SELECT 1 FROM external_events WHERE id = $1 AND user_id = $2)
	`, eventID, userID).Scan(&exists)
	if err != nil {
		return nil, err
	}
	if !exists {
		return nil, fmt.Errorf("event not found")
	}

	// Verify the card belongs to the user
	err = db.QueryRow(`
		SELECT EXISTS(SELECT 1 FROM cards WHERE id = $1 AND user_id = $2)
	`, cardPK, userID).Scan(&exists)
	if err != nil {
		return nil, err
	}
	if !exists {
		return nil, fmt.Errorf("card not found")
	}

	// Update the link
	_, err = db.Exec(`
		UPDATE external_events
		SET card_pk = $1, updated_at = NOW()
		WHERE id = $2
	`, cardPK, eventID)
	if err != nil {
		return nil, fmt.Errorf("failed to link event to card: %w", err)
	}

	// Return the updated event
	query := `
		SELECT id, user_id, external_calendar_id, title, description,
		       start_time, end_time, all_day, location,
		       external_uid, external_url, recurrence_rule,
		       recurrence_id, recurrence_instance,
		       color, card_pk, created_at, updated_at, last_synced_at
		FROM external_events
		WHERE id = $1
	`

	var event models.ExternalEvent
	var description, location, externalUID, externalURL, recurrenceRule, recurrenceID, color sql.NullString
	var externalCalendarID sql.NullInt32
	var linkedCardPK sql.NullInt32
	var recurrenceInstance sql.NullInt32
	var lastSyncedAt sql.NullTime

	err = db.QueryRow(query, eventID).Scan(
		&event.ID,
		&event.UserID,
		&externalCalendarID,
		&event.Title,
		&description,
		&event.StartTime,
		&event.EndTime,
		&event.AllDay,
		&location,
		&externalUID,
		&externalURL,
		&recurrenceRule,
		&recurrenceID,
		&recurrenceInstance,
		&color,
		&linkedCardPK,
		&event.CreatedAt,
		&event.UpdatedAt,
		&lastSyncedAt,
	)
	if err != nil {
		return nil, err
	}

	// Convert nullable fields to pointers
	if externalCalendarID.Valid {
		id := int(externalCalendarID.Int32)
		event.ExternalCalendarID = &id
	}
	if description.Valid {
		event.Description = &description.String
	}
	if location.Valid {
		event.Location = &location.String
	}
	if externalUID.Valid {
		event.ExternalUID = &externalUID.String
	}
	if externalURL.Valid {
		event.ExternalURL = &externalURL.String
	}
	if recurrenceRule.Valid {
		event.RecurrenceRule = &recurrenceRule.String
	}
	if recurrenceID.Valid {
		event.RecurrenceID = &recurrenceID.String
	}
	if recurrenceInstance.Valid {
		instance := int(recurrenceInstance.Int32)
		event.RecurrenceInstance = &instance
	}
	if color.Valid {
		event.Color = &color.String
	}
	if linkedCardPK.Valid {
		pk := int(linkedCardPK.Int32)
		event.CardPK = &pk
	}
	if lastSyncedAt.Valid {
		event.LastSyncedAt = &lastSyncedAt.Time
	}

	return &event, nil
}

// ParseISODate parses an ISO 8601 date string
func ParseISODate(dateStr string) (time.Time, error) {
	return time.Parse(time.RFC3339, dateStr)
}

// ValidateDateRange checks if start is before end
func ValidateDateRange(start, end time.Time) error {
	if start.After(end) {
		return fmt.Errorf("start date must be before end date")
	}
	return nil
}
